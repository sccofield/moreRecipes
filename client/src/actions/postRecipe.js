import axios from 'axios';
import actionTypes from './actionTypes';

const { POST_RECIPE } = actionTypes;

const token = localStorage.getItem('token');

export const postRecipeAction = recipe => ({
  type: POST_RECIPE,
  recipe
});

const postRecipeActionCreator = recipe => dispatch => axios
  .post('api/v1/recipes', recipe, { headers: { token } })
  .then((res) => {
    const recipeData = res.data.recipe;
    dispatch(postRecipeAction(recipeData));
    return Promise.resolve();
  });

export default postRecipeActionCreator;
