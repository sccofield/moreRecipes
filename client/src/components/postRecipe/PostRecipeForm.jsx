import React from 'react';
import Dropzone from 'react-dropzone';
import PropTypes from 'prop-types';


const PostRecipeForm = props => (
  <main className="container">

    <div className="row justify-content-md-center">
      <div className="col-md-8">

        <h1>Add New Recipe</h1>
        {props.state.errors &&
          <ul className="errorMessage">{
            props.state.errors.map(error => (
              <li key={Math.floor(Math.random() * 1000)
              }
              >{error}
              </li>))}
          </ul>
        }

        <form onSubmit={props.onSubmit}>
          <div className="form-group">
            <label htmlFor="recipeTitle">Recipe title</label>
            <input
              type="text"
              className="form-control"
              name="title"
              id="recipeTitle"
              placeholder="Enter recipe title"
              value={props.state.title}
              onChange={props.onChange}
              required
            />
          </div>
          <div className="form-group">
            <label htmlFor="ingredients">Recipe ingredients</label>

            <textarea
              className="form-control"
              name="ingredients"
              id="ingredients"
              placeholder="Enter recipe ingredients seperated by commas"
              rows="5"
              value={props.state.ingredients}
              onChange={props.onChange}
              required
            />
            <small id="emailHelp" className="form-text text-muted">
              seperate each ingredients with a comma (,)
            </small>

          </div>

          <div className="form-group">
            <label htmlFor="post">Recipe description</label>
            <textarea
              className="form-control"
              name="description"
              id="description"
              rows="5"
              placeholder="Enter recipe description"
              value={props.state.description}
              onChange={props.onChange}
              required
            />
          </div>

          <div className="form-group">
            <label htmlFor="imgurl">Recipe image</label>
            <Dropzone
              onDrop={props.handleDrop}
              multiple={false}
            >
              <p className="text-center pt-4">
                Click to upload image
              </p>
              { props.imagePreview &&

              <div className="form-group">
                <img
                  src={props.imagePreview}
                  alt=""
                  className="img img-responsive img-thumbnail imagePreview"
                />
              </div>
              }
            </Dropzone>


          </div>


          <button type="submit" className="btn btn-primary">Submit</button>
        </form>

      </div>
    </div>
  </main>

);

PostRecipeForm.propTypes = {
  errors: PropTypes.arrayOf(PropTypes.string).isRequired,
  handleDrop: PropTypes.func.isRequired,
  imagePreview: PropTypes.string,
  onChange: PropTypes.func.isRequired,
  onSubmit: PropTypes.func.isRequired,
  state: PropTypes.shape({
    description: PropTypes.string,
    errors: PropTypes.array,
    image: PropTypes.string,
    ingredients: PropTypes.string,
    title: PropTypes.string
  }).isRequired,
};

PostRecipeForm.defaultProps = {
  imagePreview: null
};

export default PostRecipeForm;
