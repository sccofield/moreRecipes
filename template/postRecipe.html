<!DOCTYPE html>
<html lang="en">

<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <meta http-equiv="X-UA-Compatible" content="ie=edge">
  <title>More Recipe</title>
  <link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.7.0/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-wvfXpqpZZVQGK6TAh5PVlGOfQNHSoD2xbE+QkPxCAFlNEevoEH3Sl0sibVcOQVnN"
    crossorigin="anonymous">
  <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/4.0.0-beta/css/bootstrap.min.css" integrity="sha384-/Y6pD6FV/Vv2HJnA6t+vslU6fwYXjCFtcEpHbNJ0lyAFsXTsjBbfaDjzALeQsN6M"
    crossorigin="anonymous">
  <link rel="stylesheet" href="css/style.css" />
  <link href="https://fonts.googleapis.com/css?family=Lobster" rel="stylesheet">


  <body>
    <header style="background-color: black; background-image: none;">
      <div class="container">

        <nav class="navbar navbar-dark navbar-expand-lg justify-content-between py-3 sticky">
          <button class="navbar-toggler" type="button" data-toggle="collapse" data-target=".dual-nav">
            <span class="navbar-toggler-icon"></span>
          </button>
          <div class="navbar-collapse collapse dual-nav">
            <ul class="navbar-nav nav-menu">
              <li class="nav-item">
                <a class="nav-link" href="recipes.html">Explore Recipes</a>
              </li>
              <li class="nav-item">
                <a class="nav-link" href="postRecipe.html">Post Recipe</a>
              </li>
              <li class="nav-item">
                <a class="nav-link" href="#">About Us</a>
              </li>
            </ul>
          </div>
          <div class="brand">

            <a href="index.html" class="navbar-brand mr-0">More Recipes</a>
          </div>
          <div class="navbar-collapse collapse dual-nav">
            <ul class="nav navbar-nav ml-auto">
              <li class="nav-item">
                <a class="nav-link" href="#">Login</a>
              </li>
              <li class="nav-item">
                <a class="nav-link" href="signup.html">Register</a>
              </li>
            </ul>
          </div>
        </nav>



      </div>

    </header>
    
    <main class="container">
        
            <div class="row">
              <div class="col-md-9 col-sm-9 col-xs-12">
        
              <h1>Post Recipe</h1>
        
              <form action="/addpost" method="post">
          <div class="form-group">
            <label for="recipeTitle">Recipe Title</label>
            <input type="text" class="form-control" name="title" id="recipeTitle" placeholder="Enter recipe title">
          </div>
          <div class="form-group">
            <label for="post">Enter Ingredients</label>
            <textarea class="form-control" name="content" id="ingredients" placeholder="Enter recipe ingredients"rows="5"></textarea>
          </div>
          <div class="form-group">
              <label for="post">Enter Methods</label>
              <textarea class="form-control" name="content" id="methods" rows="5" placeholder="Enter recipe method"></textarea>
            </div>
          <div class="form-group">
            <label for="tags">tags</label>
            <input type="text" class="form-control" name="tags" id="tags" placeholder="Enter tags">
          </div>
          <div class="form-group">
            <label for="imgurl">img url</label>
            <input type="text" name="imglink" class="form-control" id="imgurl" placeholder="Enter img url">
          </div>
        
          <button type="submit" class="btn btn-primary">Submit</button>
        </form>
      </main>



    <footer>
      <div class="container ">

        <div class="row">
          <div class="col-md-3">
            <div class="brand">
              <a href="#">More Recipes</a>
            </div>

          </div>

          <div class="col-md-3">
            <h3>Latest Recipes</h3>
            <ul class="">
              <li>lorem ispum lorem</li>
              <li>lorem ispum lorem</li>
              <li>lorem ispum lorem</li>

            </ul>

          </div>

          <div class="col-md-3">
            <h3>Links</h3>
            <ul>
              <li>About us</li>
              <li>FAQ</li>
              <li>Polices</li>
              <li>Contact Us</li>

            </ul>

          </div>
          <div class="col-md-3">
            <h3>Contact us</h3>
            <p>Andela Epic Towers
              <br/> Ikorodu, Lagos
              <br/> Nigeria.
              <br/>

          </div>

        </div>
      </div>
    </footer>





    <script src="https://code.jquery.com/jquery-3.2.1.slim.min.js" integrity="sha384-KJ3o2DKtIkvYIK3UENzmM7KCkRr/rE9/Qpg6aAZGJwFDMVNA/GpGFF93hXpG5KkN"
      crossorigin="anonymous"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/popper.js/1.11.0/umd/popper.min.js" integrity="sha384-b/U6ypiBEHpOf/4+1nzFpr53nxSS+GLCkfwBdFNTxtclqqenISfwAzpKaMNFNmj4"
      crossorigin="anonymous"></script>
    <script src="https://maxcdn.bootstrapcdn.com/bootstrap/4.0.0-beta/js/bootstrap.min.js" integrity="sha384-h0AbiXch4ZDo7tp9hKZ4TsHbi047NrKGLO3SEJAg45jXxnGIfYzk4Si90RDIqNm1"
      crossorigin="anonymous"></script>

  </body>

</html>