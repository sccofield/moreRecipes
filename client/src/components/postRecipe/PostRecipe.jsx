import React from 'react';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import toastr from 'toastr';
import PropTypes from 'prop-types';
import axios from 'axios';
import PageHeader from '../PageHeader';
import Footer from '../Footer';
import PostRecipeForm from './PostRecipeForm';
import FormValidator from './FormValidator';
import postRecipeActionCreator,
{ clearNewRecipe } from '../../actions/postRecipe';

/**
 *
 *
 * @class PostRecipe
 * @extends {React.Component}
 */
class PostRecipe extends React.Component {
  /**
   * Creates an instance of PostRecipe.
   * @memberof PostRecipe
   */
  constructor() {
    super();

    this.state = {
      title: '',
      description: '',
      ingredients: '',
      errors: [],
      image: null
    };
    this.handleDrop = this.handleDrop.bind(this);
    this.onChange = this.onChange.bind(this);
    this.onSubmit = this.onSubmit.bind(this);
  }

  /**
   *
   * @returns {obj} obj
   * @memberof PostRecipe
   */
  componentWillMount() {
    this.props.clearNewRecipe();
  }

  /**
 *
 * @returns {obj} obj
 * @param {any} nextProps
 * @memberof PostRecipe
 */
  componentWillReceiveProps(nextProps) {
    if (nextProps.newRecipe) {
      toastr.success('Your recipe has been added');
      this.props.history.push(`recipes/${nextProps.newRecipe.id}`);
    }
  }

  /**
   * @description handles form change events
   * @returns {null} null
   * @param {any} event
   * @memberof Login
   */
  onChange(event) {
    this.setState({ [event.target.name]: event.target.value });
  }

  /**
   * @description handles form submit events
   * @returns {null} null
   * @param {any} event
   * @memberof Login
   */
  onSubmit(event) {
    event.preventDefault();
    if (this.validate()) {
      const imageUpload = new FormData();

      if (this.state.image) {
        imageUpload.append('file', this.state.image);
        imageUpload.append('upload_preset', 'ftpwkq8d');
        imageUpload.append('api_key', '897164162315899');
        axios.post(
          'https://api.cloudinary.com/v1_1/sccofield/image/upload',
          imageUpload
        )
          .then((res) => {
            const postData = this.state;
            postData.imageUrl = res.data.secure_url;
            this.props.postRecipeActionCreator(postData);
          });
      } else {
        this.props.postRecipeActionCreator(this.state);
      }
      //
    }
  }

  /**
 *
 * @returns {obj} object
 * @param {any} files
 * @memberof PostRecipe
 */
  handleDrop(files) {
    this.setState({ image: files[0] });
  }

  /**
 *
 *
 * @returns {boolean} boolean
 * @memberof Login
 */
  validate() {
    const { errors } = FormValidator(this.state);
    if (errors.length > 0) {
      this.setState({ errors });
      return false;
    }
    this.setState({ errors: [] });
    return true;
  }

  /**
   *
   *
   * @returns {null} null
   * @memberof PostRecipe
   */
  render() {
    return (
      <div className="main">
        <PageHeader />
        <div className="container">
          <div className="row content">
            <div className="col-lg-10">
              <PostRecipeForm
                className="mb-4"
                onChange={this.onChange}
                state={this.state}
                onSubmit={this.onSubmit}
                errors={this.state.errors}
                errorMessage={this.props.errorMessage}
                handleDrop={this.handleDrop}
                imagePreview={this.state.image && this.state.image.preview}
                handleEditorChange={this.handleEditorChange}
              />
            </div>
          </div>
        </div>
        <Footer />
      </div>
    );
  }
}

const mapStateToProps = state => ({
  newRecipe: state.recipe.newRecipe
});

const mapDispatchToProps = dispatch =>
  bindActionCreators({ postRecipeActionCreator, clearNewRecipe }, dispatch);

PostRecipe.propTypes = {
  clearNewRecipe: PropTypes.func.isRequired,
  history: PropTypes.shape({
    push: PropTypes.func
  }).isRequired,
  postRecipeActionCreator: PropTypes.func.isRequired,
  newRecipe: PropTypes.shape({
    id: PropTypes.string
  }),
  errorMessage: PropTypes.arrayOf(PropTypes.string)
};

PostRecipe.defaultProps = {
  newRecipe: null,
  errorMessage: null
};

export default connect(mapStateToProps, mapDispatchToProps)(PostRecipe);
