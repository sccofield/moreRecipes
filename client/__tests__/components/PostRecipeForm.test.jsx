import React from 'react';
import toJSON from 'enzyme-to-json';
import sinon from 'sinon';
import PostRecipeForm from '../../src/components/postRecipe/PostRecipeForm';
import { PostRecipe } from '../../src/components/postRecipe/PostRecipe';

describe('Recipe component', () => {
  const props = {
    state: {
      description: 'description',
      image: 'image',
      ingredients: 'ingredients',
      title: 'title'
    }
  };

  const errorProps = {
    state: {
      description: 'description',
      image: 'image',
      ingredients: 'ingredients',
      title: 'title',
      errors: ['error']
    }
  };

  const imageProps = {
    state: {
      description: 'description',
      image: 'image',
      ingredients: 'ingredients',
      title: 'title',
    },
    imagePreview: 'image'
  };


  it('should render postRecipe form when props is given', () => {
    const wrapper = shallow(<PostRecipeForm {...props} />);
    expect(wrapper.find('div').length).toEqual(6);
    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  it('should show error message when there is an error in state', () => {
    const wrapper = shallow(<PostRecipeForm {...errorProps} />);
    expect(wrapper.find('.errorMessage').text()).toBe('error');
    expect(wrapper.find('.errorMessage li').length).toEqual(1);
  });

  it('should show the image if it has been uploaded', () => {
    const wrapper = shallow(<PostRecipeForm {...imageProps} />);
    expect(wrapper.find('.imagePreview').length).toEqual(1);
  });

  it('should call the onsubmit function when the form is submitted', () => {
    const spy = sinon.spy(PostRecipe.prototype, 'onSubmit');
    const Props = {
      ...props,
      onSubmit: spy
    };
    const wrapper = shallow(<PostRecipeForm {...Props} />);
    try {
      wrapper.find('form').simulate('submit', { preventDefault: jest.fn() });
    } catch (error) {

    }
    expect(spy.called).toBeTruthy();
  });
});

