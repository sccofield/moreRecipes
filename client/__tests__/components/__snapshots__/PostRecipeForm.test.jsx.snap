// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Recipe component should render postRecipe form when props is given 1`] = `
<main
  className="container"
>
  <div
    className="row justify-content-md-center"
  >
    <div
      className="col-md-8"
    >
      <h1>
        Add New Recipe
      </h1>
      <form
        onSubmit={null}
      >
        <div
          className="form-group"
        >
          <label
            htmlFor="recipeTitle"
          >
            Recipe title
          </label>
          <input
            className="form-control"
            id="recipeTitle"
            name="title"
            onChange={null}
            placeholder="Enter recipe title"
            required={true}
            type="text"
            value="title"
          />
        </div>
        <div
          className="form-group"
        >
          <label
            htmlFor="ingredients"
          >
            Recipe ingredients
          </label>
          <textarea
            className="form-control"
            id="ingredients"
            name="ingredients"
            onChange={null}
            placeholder="Enter recipe ingredients seperated by commas"
            required={true}
            rows="5"
            value="ingredients"
          />
          <small
            className="form-text text-muted"
            id="emailHelp"
          >
            seperate each ingredients with a comma (,)
          </small>
        </div>
        <div
          className="form-group"
        >
          <label
            htmlFor="post"
          >
            Recipe description
          </label>
          <textarea
            className="form-control"
            id="description"
            name="description"
            onChange={null}
            placeholder="Enter recipe description"
            required={true}
            rows="5"
            value="description"
          />
        </div>
        <div
          className="form-group"
        >
          <label
            htmlFor="imgurl"
          >
            Recipe image
          </label>
          <t
            disableClick={false}
            disablePreview={false}
            disabled={false}
            maxSize={Infinity}
            minSize={0}
            multiple={false}
            onDrop={null}
            preventDropOnDocument={true}
          >
            <p
              className="text-center pt-4"
            >
              Click to upload image
            </p>
          </t>
        </div>
        <button
          className="btn btn-primary"
          id="recipeSubmit"
          type="submit"
        >
          Submit
        </button>
      </form>
    </div>
  </div>
</main>
`;
